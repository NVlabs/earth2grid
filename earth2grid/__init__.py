# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import torch

from earth2grid import base, healpix, latlon
from earth2grid._regrid import (
    BilinearInterpolator,
    Identity,
    S2NearestNeighborInterpolator,
)

__all__ = [
    "base",
    "healpix",
    "latlon",
    "get_regridder",
    "BilinearInterpolator",
    "S2NearestNeighborInterpolator",
    "S2LinearBarycentricInterpolator",
]


def get_regridder(src: base.Grid, dest: base.Grid) -> torch.nn.Module:
    """Get a regridder from `src` to `dest`"""
    if src == dest:
        return Identity()
    elif isinstance(src, latlon.LatLonGrid) and isinstance(dest, latlon.LatLonGrid):
        return src.get_bilinear_regridder_to(dest.lat, dest.lon)
    elif isinstance(src, latlon.LatLonGrid) and isinstance(dest, healpix.Grid):
        return src.get_bilinear_regridder_to(dest.lat, dest.lon)
    elif isinstance(src, healpix.Grid):
        return src.get_bilinear_regridder_to(dest.lat, dest.lon)
    elif isinstance(dest, healpix.Grid):
        return src.get_healpix_regridder(dest)  # type: ignore

    raise ValueError(src, dest, "not supported.")
