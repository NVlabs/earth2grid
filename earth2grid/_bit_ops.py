# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


def compact_bits(bits):
    # Remove interleaved 0 bits
    bits = bits & 0x5555555555555555  # Mask: 01010101...
    # example implementation for 1 byte
    # 0a0b0c0d
    # 00ab00cd # (x | x >> 1)  & 00110011 = 0x33
    # 0000abcd # (x | x >> 2)  & 00001111 = 0x0F
    # --------
    # abc0d
    bits = (bits | (bits >> 1)) & 0x3333333333333333  # noqa
    bits = (bits | (bits >> 2)) & 0x0F0F0F0F0F0F0F0F  # noqa
    bits = (bits | (bits >> 4)) & 0x00FF00FF00FF00FF  # noqa
    bits = (bits | (bits >> 8)) & 0x0000FFFF0000FFFF  # noqa
    bits = (bits | (bits >> 16)) & 0x00000000FFFFFFFF  # noqa
    return bits


def spread_bits(bits):
    """
    bits is a 32 bit number (stored in int64)
    algorithm starts by moving the first 16 bits to the left by 16
    and proceeding recursively
    """
    # example implementation for a byte
    # 0000abcd
    # 00ab00cd # (x | x <<2)  & 00110011 = 0x33
    # 0a0b0c0d # (x | x <<1)  & 01010100 = 0x55
    # --------
    # abc0d
    bits = (bits | (bits << 16)) & 0x0000FFFF0000FFFF  # noqa
    bits = (bits | (bits << 8)) & 0x00FF00FF00FF00FF  # noqa
    bits = (bits | (bits << 4)) & 0x0F0F0F0F0F0F0F0F  # noqa
    bits = (bits | (bits << 2)) & 0x3333333333333333  # noqa
    bits = (bits | (bits << 1)) & 0x5555555555555555  # noqa
    return bits
