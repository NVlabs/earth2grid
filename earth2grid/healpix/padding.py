# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.import torch
import torch


def local2xy(
    nside: int, x: torch.Tensor, y: torch.Tensor, face: torch.Tensor, right_first: bool = True
) -> torch.Tensor:
    """Convert a local x, y coordinate in a given face (S origin) to a global pixel index

    The local coordinates can be < 0 or > nside

    This can be used to implement padding

    Args:
        nside: int
        x: local coordinate [-nside, 2 * nside), origin=S
        y: local coordinate [-nside, 2 * nside), origin=S
        face: index of the face [0, 12)
        right_first: if True then traverse to the face in the x-direction first

    Returns:
        global XY pixel index
    """
    # adjacency graph (8 neighbors, counter-clockwise from S)
    # Any faces > 11 are missing
    neighbors = torch.tensor(
        [
            # pole
            [8, 5, 1, 1, 2, 3, 3, 4],
            [9, 6, 2, 2, 3, 0, 0, 5],
            [10, 7, 3, 3, 0, 1, 1, 6],
            [11, 4, 0, 0, 1, 2, 2, 7],
            # equator
            [16, 8, 5, 0, 12, 3, 7, 11],
            [17, 9, 6, 1, 13, 0, 4, 8],
            [18, 10, 7, 2, 14, 1, 5, 9],
            [19, 11, 4, 3, 15, 2, 6, 10],
            # south pole
            [10, 9, 9, 5, 0, 4, 11, 11],
            [11, 10, 10, 6, 1, 5, 8, 8],
            [8, 11, 11, 7, 2, 6, 9, 9],
            [9, 8, 8, 4, 3, 7, 10, 10],
        ],
        device=x.device,
    )

    # number of left turns the path takes while traversing from face i to j
    turns = torch.tensor(
        [
            # pole
            [0, 0, 0, 3, 2, 1, 0, 0],
            # equator
            [0, 0, 0, 0, 0, 0, 0, 0],
            # south pole
            [2, 1, 0, 0, 0, 0, 0, 3],
        ],
        device=x.device,
    )
    # x direction
    face_shift_x = x // nside
    face_shift_y = y // nside

    # TODO what if more face_shift_x, face_shift_y = 2, 1 or similar?
    # which direction should we traverse faces in?
    direction_lookup = torch.tensor([[0, 7, 6], [1, -1, 5], [2, 3, 4]], device=x.device)

    direction = direction_lookup[face_shift_x + 1, face_shift_y + 1]
    new_face = torch.where(direction != -1, neighbors[face, direction], face)
    origin = torch.where(direction != -1, turns[face // 4, direction], 0)

    # rotate back to origin = S convection
    for i in range(1, 4):
        nx, ny = _rotate(i, x, y)
        x = torch.where(origin == i, nx, x)
        y = torch.where(origin == i, ny, y)

    face = new_face
    return x % nside, y % nside, face
    return torch.where(face != -1, face * (nside * nside) + (y % nside) * nside + (x % nside), -1)


def _rotate(rotations: int, x, y):
    """rotate (x,y) counter clockwise"""
    k = rotations % 4
    # Apply the rotation based on k
    if k == 1:  # 90 degrees counterclockwise
        return -y - 1, x
    elif k == 2:  # 180 degrees
        return -x - 1, -y - 1
    elif k == 3:  # 270 degrees counterclockwise
        return y, -x - 1
    else:  # k == 0, no change
        return x, y


def _get_xy(nside, f, x, y):
    return torch.where(f < 12, f * nside**2 + y * nside + x, -1)


def _xy_with_filled_tile(nside, x1, y1, f1):
    """Handles an points with missing tile information following the HPXPAD strategy

    Missing tiles are defined for face >= 12. 12-16 are the N missing tiles, and
    16-20 the south missing tiles (from W to east).

    Since there is an ambiguity return both x and y.
    """

    # handle missing tiles
    # for N tiles
    # f(x, y) is filled by shuffling from the left
    # case x > y: (x, y) -> (y, )
    # examples  (for nside = 4)
    #   (3, 1)-> (0, 1)
    #   (3, 2) -> (1, 2)
    #   (3, 3) -> (2, 3)
    # generalize
    #   (i, j)-> (i + j, j)  in the missing face
    #   (i' - j, j) -> (i', j)

    is_missing_n_pole_tile = (f1 >= 12) & (f1 < 16)
    west_face = torch.where(is_missing_n_pole_tile, f1 - 13, 0) % 4
    east_face = (west_face + 1) % 4

    # two sets of indices
    def _pad_from_west(x1, y1, west_face):
        f_west = torch.where(is_missing_n_pole_tile & (x1 <= y1), west_face, f1)
        x_west = torch.where(is_missing_n_pole_tile & (x1 < y1), (x1 - y1) % nside, x1)
        x_west = torch.where(is_missing_n_pole_tile & (x1 == y1), nside - 1, x_west)
        y_west = y1
        return x_west, y_west, f_west

    x_west, y_west, f_west = _pad_from_west(x1, y1, west_face)
    y_east, x_east, f_east = _pad_from_west(y1, x1, east_face)

    # S pole
    is_missing_s_pole_tile = (f1 >= 16) & (f1 < 20)
    east_face = (f1 - 16) % 4 + 8
    west_face = (east_face - 9) % 4 + 8

    # two sets of indices
    def _pad_from_east(x1, y1, east_face, f1):
        """Test cases

        (1, 0) -> (0, 0)
        (3, 2) -> (0, 2)
        """
        f_west = torch.where(is_missing_s_pole_tile & (x1 >= y1), east_face, f1)
        # x_west = torch.where(is_missing_s_pole_tile & (x1 > y1), 1(x1-y1) %nside, x1)
        x_west = torch.where(is_missing_s_pole_tile & (x1 > y1), (x1 - y1 - 1) % nside, x1)
        x_west = torch.where(is_missing_s_pole_tile & (x1 == y1), 0, x_west)
        y_west = y1
        return x_west, y_west, f_west

    x_west, y_west, f_west = _pad_from_east(x_west, y_west, east_face, f_west)
    y_east, x_east, f_east = _pad_from_east(y_east, x_east, west_face, f_east)

    xy_west = _get_xy(nside, f_west, x_west, y_west)
    xy_east = _get_xy(nside, f_east, x_east, y_east)

    return xy_west, xy_east


def pad(x, pad):
    # x - (n, c, f, x, y) in origin=S order
    n, c = x.shape[:2]
    nside = x.shape[-1]
    x = x.reshape((n, c, -1))

    # setup padded grid
    i = torch.arange(-pad, nside + pad, device=x.device)
    j = torch.arange(-pad, nside + pad, device=x.device)
    f = torch.arange(12, device=x.device)

    # get indices in source data for target points
    f, j, i = torch.meshgrid(f, j, i, indexing="ij")
    i1, j1, f1 = local2xy(nside, i, j, f)
    xy_east, xy_west = _xy_with_filled_tile(nside, i1, j1, f1)

    # average the potential ambiguous regions
    padded_from_west = torch.where(xy_west >= 0, x[..., xy_west], 0)
    padded_from_east = torch.where(xy_east >= 0, x[..., xy_east], 0)
    denom = (xy_west >= 0).int() + (xy_east >= 0).int()

    return (padded_from_east + padded_from_west) / denom
