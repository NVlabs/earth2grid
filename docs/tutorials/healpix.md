---
file_format: mystnb
kernelspec:
  name: python3
---

# HEALPix Tutorial

This tutorial demonstrates key HealPIX functionality including:

1. Visualization using double pixelization
2. Zonal averaging
3. Regridding between different grids
4. Reordering between different pixel ordering schemes
5. Padding for machine learning applications

## Installation

Earth2grid has some compiled components (in CUDA and C). So some care is required to install. First, install pytorch, and setuptools

```
# python3 -m venv .venv # make venv
# source .venv/bin/activate # activate venv
pip install torch setuptools
pip install --no-build-isolation https://github.com/NVlabs/earth2grid/archive/main.tar.gz
```
:::{warning}
It is very important to use the `--no-build-isolation` flag. Otherwise, the library may be built against a different version of pytorch than you have installed, this will cause strange "missing symbols" errors.
:::


## Setup

First, let's import the necessary libraries and create a sample HealPIX grid.

```{code-cell}
#| label: setup
import matplotlib.pyplot as plt
import numpy as np
import torch

import earth2grid
from earth2grid import healpix

# Create a sample HealPIX grid
level = 3  # Resolution level (higher = finer resolution)
grid = healpix.Grid(level=level, pixel_order=healpix.PixelOrder.RING)
print(grid.lat[:5])
```


```{code-cell}
:tags: [hide-input]

SHOW_AXES = False

def label_pixels(z):
    z = torch.as_tensor(z)
    npix = z.shape[0]
    i_ring = torch.arange(npix)
    nside = healpix.npix2nside(npix)
    y, x =healpix.ring2double(nside, i_ring)
    if label_pixels:
        for i in range(npix):
            plt.text((x[i] + nside).item() / 2, nside-(y[i]).item() / 2, z[i].item(), ha='center', va='center', fontsize=12, color='white', clip_on=True)

```

## Visualization using Double Pixelization

Double pixelization provides a visually appealing way to view HealPIX data without interpolation, preserving the native pixel structure. This is particularly useful for quick visualization with image viewers without distorting the native pixels of the image.

```{code-cell}
# Convert to double pixelization
field_double = healpix.to_double_pixelization(grid.lat)

plt.figure(figsize=(12, 4))
plt.imshow(field_double)
plt.colorbar(label='Field Value')
plt.title('Double Pixelization Visualization')
plt.show()
```

## Zonal Averaging

Zonal averaging computes the mean value of a field along latitude circles. This is useful for analyzing latitudinal patterns in the data.

```{code-cell}
# Compute the zonal average of the field
field = np.cos(np.deg2rad(grid.lat))
zonal_avg = healpix.zonal_average(field)
lat = healpix.zonal_average(grid.lat)

```

```{code-cell}
:tags: [hide-input]
# Plot the zonal average
plt.figure(figsize=(8, 4))
plt.plot(lat, zonal_avg)
plt.xlabel('Latitude (degrees)')
plt.ylabel('Zonal Average')
plt.title('Zonal Average of Sample Field')
plt.grid(True)
plt.show()
```

## Regridding

Regridding allows us to transform data between different grid types. Here we demonstrate two types of regridding:
1. From HEALPix to a regular lat-lon grid
2. Between HEALPix grids of different resolutions

### Regridding to Lat-Lon

`earth2grid` has builtin utilities for bilinear interpolation from any supported grid to arbitrary lat-lon points.

#### HEALPIX to Lat-Lon


```{code-cell}
#| label: regrid-latlon
# Create a lat-lon grid
nlat, nlon = 33, 64
latlon_grid = earth2grid.latlon.equiangular_lat_lon_grid(nlat, nlon)

# Create regridder from HealPIX to lat-lon
regridder = earth2grid.get_regridder(grid, latlon_grid)

# Regrid the field
# must be a torch tensor
field = np.cos(np.deg2rad(grid.lat+ 40)) ** 2 + 0.5 * np.sin(np.deg2rad(grid.lon))
field_regridded = regridder(torch.from_numpy(field))
```


```{code-cell}
:tags: [hide-input]
# Plot original and regridded fields
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 4))

# Original HealPIX field
plt.sca(ax1)
im1 = healpix.pcolormesh(field)
ax1.set_title('Original HealPIX Field')
plt.colorbar(im1, ax=ax1)

# Regridded field
im2 = ax2.pcolormesh(latlon_grid.lon, latlon_grid.lat, field_regridded)
ax2.set_title('Regridded to Lat-Lon')
plt.colorbar(im2, ax=ax2)

plt.tight_layout()
plt.show()
```

A more verbose way, but flexible, way to do this is using the `.get_bilinear_regridder_to` method.

```{code-cell}
#| label: regrid-latlon-verbose

lat = np.linspace(-90, 90, 33)
lon = np.linspace(0, 360, 64)

regridder = grid.get_bilinear_regridder_to(lat[:, np.newaxis], lon[np.newaxis, :])
field_regridded = regridder(torch.from_numpy(field))
print(field_regridded.shape)
```

:::{note}
Exercise: use `get_bilinear_regridder_to` to regrid to a list of unstructured points.
:::

The regridding objects (and most operations in earth2grid) are fully differentiable and GPU capable
```{code-cell}

x = torch.randn(grid.shape)
x.requires_grad_(True)
regridder.to(x.dtype) # need to make sure the regridder is the same type as the input
y = regridder(x)
(y ** 2).sum().backward()
print(x.grad[:5])
```


### Lat-Lon to HEALPIX

```{code-cell}
from earth2grid import latlon
llgrid = latlon.LatLonGrid(lat=lat, lon=lon)
hpxgrid = healpix.Grid(level=6)
regridder = llgrid.get_bilinear_regridder_to(hpxgrid.lon, hpxgrid.lat)

field = lat[:, None] + 0 * lon # trick to get (nlat, nlon) array
out = regridder(torch.from_numpy(field))
print(out.shape)
```

## HEALPix Pixel Orders

earth2grid supports three main pixel ordering schemes:

- RING: Pixels are ordered in rings of constant latitude
- NEST: Pixels are ordered in a hierarchical, nested pattern
- XY: Pixels are ordered in a Cartesian-like grid within each face, with configurable origin and direction

The ability to reorder between these schemes is crucial for compatibility with different HEALPix implementations and for certain operations.

This is the default RING order:
```{code-cell}
:tags: [hide-input]

def show_pixel_order(nside, pixel_order):
    npix = 12 * nside ** 2
    i = torch.arange(npix)
    ir = healpix.reorder(i, pixel_order, healpix.PixelOrder.RING)
    healpix.pcolormesh(ir)
    label_pixels(ir)
    if isinstance(pixel_order, healpix.XY):
        clockwise = pixel_order.clockwise
        plt.title(f"{pixel_order.origin.name}, {clockwise=}")

show_pixel_order(2, healpix.PixelOrder.RING)
```

```{code-cell}
:tags: [hide-input]
show_pixel_order(1, healpix.PixelOrder.RING)
```


And the NEST
```{code-cell}
:tags: [hide-input]
show_pixel_order(2, healpix.PixelOrder.NEST)
```

While NEST and RING are classic, we often want to treat each base pixel of the healpix grid as a 2d array. We have a XY pixel ordering object that allows us to do this.

```{code-cell}
:tags: [hide-input]
def show_pixel_order(nside, pixel_order):
    npix = 12 * nside ** 2
    i = torch.arange(npix)
    ir = healpix.reorder(i, pixel_order, healpix.PixelOrder.RING)
    healpix.pcolormesh(ir.clamp(0, nside**2))
    label_pixels(ir)

    if isinstance(pixel_order, healpix.XY):
        clockwise = pixel_order.clockwise
        plt.title(f"XY(origin={pixel_order.origin.name}, {clockwise=})")
    else:
        plt.title(f"{pixel_order.name}")

nrow = 2
ncol = 2
plt.subplot(nrow, ncol, 1)
show_pixel_order(4, healpix.PixelOrder.NEST)
plt.ylim(0, 4)
plt.xlim(2, 6)

plt.subplot(nrow, ncol, 2)
show_pixel_order(4, healpix.XY())
plt.ylim(0, 4)
plt.xlim(2, 6)

plt.subplot(nrow, ncol, 3)
show_pixel_order(4, healpix.XY(origin=healpix.Compass.N, clockwise=False))
plt.ylim(0, 4)
plt.xlim(2, 6)

plt.subplot(nrow, ncol, 4)
show_pixel_order(4, healpix.XY(origin=healpix.Compass.N, clockwise=True))
plt.ylim(0, 4)
plt.xlim(2, 6)
```


Why can convert between any two pixel orders like this:
```
healpix.reorder(x, healpix.PixelOrder.RING, healpix.PixelOrder.NEST) # ring to nest
healpix.reorder(x, healpix.PixelOrder.NEST, healpix.PixelOrder.XY()) # ring to XY
```



### Padding and Convolutions

The XY pixel order is useful for image analysis since it is a Cartesian-like grid. For example this is what the data looks like on each face of the sphere:

```{code-cell}
:tags: [hide-input]
level = 6
nside = 2 ** level

grid = healpix.Grid(level)
field = np.cos(np.deg2rad(grid.lat+ 40)) ** 2 + 0.5 * np.sin(np.deg2rad(grid.lon))
field = torch.from_numpy(field)

out = healpix.reorder(field, healpix.PixelOrder.RING, healpix.XY())
faces = out.reshape([12, nside, nside])
for i in range(12):
    plt.subplot(3, 4, i+1)
    plt.pcolormesh(faces[i])
    plt.gca().set_aspect('equal')
    plt.title(f"face {i}")
plt.tight_layout()
```

Padding is a basic primitive for many machine learning methods. For example, convolutions un the sphere can be implemented as a padding followed by a convolution. Because of this we have written a CUDA padding routine that works for the HEALPix grid. It requires (origin=N, orientation=clockwise) i.e. [N, E, S, W].  For convenience we have an alias for this `healpix.HEALPIX_PAD_XY`. Let's now call the padding routine and plot the padded region.

```{code-cell}
nside = 8
pad_size = 4

# Let's assume some data on the NEST grid (like is common for datasets)
grid = healpix.Grid(level=healpix.nside2level(nside), pixel_order=healpix.PixelOrder.NEST)
field = np.cos(np.deg2rad(grid.lat+ 40)) ** 2 + 0.5 * np.sin(np.deg2rad(grid.lon))
field = torch.from_numpy(field)

# now reorder, reshape and pad
out = healpix.reorder(field, grid.pixel_order, healpix.HEALPIX_PAD_XY)
z = out.reshape([1, 12, nside, nside]) # healpix.pad requires a 4d input
z_padded = healpix.pad(z, pad_size)
```

```{code-cell}
:tags: [hide-input]
# show the padded region
import matplotlib.patches as patches
plt.pcolormesh(z_padded[0, 0])
rect = patches.Rectangle((pad_size, pad_size), nside, nside, linewidth=2, edgecolor='r', facecolor='none')
plt.gca().add_patch(rect)
plt.gca().set_aspect('equal')
plt.title("Pixel 0 w/ padding");
```

To run convolutions on the sphere
```{code-cell}
# convolution 3d is used to skip over the face dimension
conv = torch.nn.Conv3d(
  in_channels=1, out_channels=1, kernel_size=[1, 3, 3], padding=0
)

pad_size = 1
b, c, f, n, n = 1, 1, 12, nside, nside


x = out.view(b, c, f, n, n).float() # some input data

# start processing
y = x
y = y.view(b * c, f, n, n) # combine (b, c) dims, healpix.pad requires a 4d input
y = healpix.pad(y, pad_size)
y = y.view(b, c, f, n+2 * pad_size, n+2 * pad_size)
y = conv(y)
print(x.shape, y.shape)

# reshape back for visualization
y = healpix.reorder(y.view(b, c, f * n  * n), healpix.HEALPIX_PAD_XY, healpix.PixelOrder.RING)
healpix.pcolormesh(y[0,0].detach())
```


## Summary

In this tutorial, we've covered several key aspects of working with HEALPix data:

1. **Visualization**: Using double pixelization to create visually appealing representations of HEALPix data while preserving the native pixel structure.

2. **Zonal Averaging**: Computing latitudinal averages to analyze patterns in the data.

3. **Regridding**: Transforming data between different grid types, including:
   - HEALPix to regular lat-lon grid
   - HEALPix to higher resolution HEALPix grid

4. **Reordering**: Converting between different HEALPix pixel ordering schemes:
   - RING: Pixels ordered in rings of constant latitude
   - NEST: Pixels ordered in a hierarchical, nested pattern
   - XY: Pixels ordered in a Cartesian-like grid within each face, with configurable origin and direction

5. **Padding**: Implementing efficient CUDA-based padding operations for machine learning applications, particularly useful for spherical convolutions.

These tools provide a comprehensive set of operations for working with global data on the sphere, making it easier to analyze, visualize, and process geophysical data for various applications including machine learning.
