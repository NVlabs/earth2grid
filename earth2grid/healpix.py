"""

From this notebook: https://colab.research.google.com/drive/1MzTyeNFiy-7RNY6UtGKsmDavX5dk6epU


Healpy has two indexing conventions NEST and RING. But for convolutions we want
2D array indexing in row or column major order. Here are some vectorized
routines `nest2xy` and `x2nest` for going in between these conventions. The
previous code shared by Dale used string computations to handle these
operations, which was probably quite slow. Here we use vectorized bit-shifting.
"""
from dataclasses import dataclass
from enum import Enum

import einops
import healpy
import numpy as np

try:
    import pyvista as pv
except ImportError:
    pv = None

from earth2grid import base


class PixelOrder(Enum):
    RING = 0
    NEST = 1
    XY = 2


@dataclass
class Grid(base.Grid):
    """A Healpix Grid"""
    level: int
    pixel_order: PixelOrder = PixelOrder.RING

    def _nside(self):
        return 2**self.level

    def _npix(self):
        return self._nside() ** 2 * 12

    def _nest_ipix(self):
        """convert to nested index number"""
        i = np.arange(self._npix())
        if self.pixel_order == PixelOrder.RING:
            return healpy.ring2nest(self._nside(), i)
        elif self.pixel_order == PixelOrder.NEST:
            return i
        else:
            return xy2nest(self._nside(), i)

    def _reorder_to_nest(self, map):
        """convert to nested index number"""
        i = np.arange(self._npix())
        if self.pixel_order == PixelOrder.RING:
            j = healpy.nest2ring(self._nside(), i)
            return map[j]
        elif self.pixel_order == PixelOrder.NEST:
            return map
        else:
            j = nest2xy(self._nside(), i)
            return map[j]

    @property
    def lat(self):
        _, lat = healpy.pix2ang(self._nside(), self._nest_ipix(), lonlat=True, nest=True)
        return lat

    @property
    def lon(self):
        lon, _ = healpy.pix2ang(self._nside(), self._nest_ipix(), lonlat=True, nest=True)
        return lon

    @property
    def shape(self) -> tuple[int, ...]:
        return (self._npix(),)

    def visualize(self, map):
        healpy.mollview(self._reorder_to_nest(map), nest=True)

    def to_pyvista(self):
        if pv is None:
            raise ImportError("Need to install pyvista")

        # Make grid
        nside = 2**self.level
        pix = self._nest_ipix()
        points = healpy.boundaries(nside, pix, step=1, nest=True)
        out = einops.rearrange(points, "n d s -> (n s) d")
        unique_points, inverse = np.unique(out, return_inverse=True, axis=0)
        assert unique_points.ndim == 2
        assert unique_points.shape[1] == 3
        inverse = einops.rearrange(inverse, "(n s) -> n s", n=pix.size)
        n, s = inverse.shape
        cells = np.ones_like(inverse, shape=(n, s + 1))
        cells[:, 0] = s
        cells[:, 1:] = inverse
        celltypes = np.full(shape=(n,), fill_value=pv.CellType.QUAD)
        grid = pv.UnstructuredGrid(cells, celltypes, unique_points)
        return grid


# nside = 2^ZOOM_LEVELS
ZOOM_LEVELS = 12


def _extract_every_other_bit(binary_number):
    result = 0
    shift_count = 0

    for i in range(ZOOM_LEVELS):
        # Check if the least significant bit is 1
        # Set the corresponding bit in the result
        result |= (binary_number & 1) << shift_count

        # Shift to the next bit to check
        binary_number = binary_number >> 2
        shift_count += 1

    return result


def nest2xy(nside, i):
    """convert NEST to XY index"""
    tile = i // nside**2
    j = i % (nside**2)
    x = _extract_every_other_bit(j)
    y = _extract_every_other_bit(j >> 1)
    return tile * nside**2 + y * nside + x


def xy2nest(nside, i):
    """convert XY index to NEST"""
    tile = i // (nside**2)
    y = (i % (nside**2)) // nside
    x = i % nside

    result = 0
    for i in range(ZOOM_LEVELS):
        # Extract the ith bit from the number
        extracted_bit = (x >> i) & 1
        result |= extracted_bit << (2 * i)

        extracted_bit = (y >> i) & 1
        result |= extracted_bit << (2 * i + 1)
    return result | (tile * nside**2)
