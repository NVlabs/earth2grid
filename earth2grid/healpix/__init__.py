# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""

HEALPix
=======

From this notebook: https://colab.research.google.com/drive/1MzTyeNFiy-7RNY6UtGKsmDavX5dk6epU



"""


from earth2grid.healpix._padding import PaddingBackends, pad, pad_backend, pad_with_dim
from earth2grid.healpix.coordinates import (
    angular_to_global,
    global_to_angular,
    global_to_face,
)
from earth2grid.healpix.core import (
    HEALPIX_PAD_XY,
    XY,
    Compass,
    Grid,
    PixelOrder,
    local2xy,
    nest2xy,
    npix2level,
    npix2nside,
    nside2level,
    reorder,
    ring2double,
    ring2xy,
    to_double_pixelization,
    to_rotated_pixelization,
    xy2xy,
    zonal_average,
)
from earth2grid.healpix.nn import conv2d
from earth2grid.healpix.visualization import pcolormesh

__all__ = [
    "angular_to_global",
    "Compass",
    "conv2d",
    "global_to_angular",
    "global_to_face",
    "Grid",
    "HEALPIX_PAD_XY",
    "local2xy",
    "nest2xy",
    "npix2level",
    "npix2nside",
    "nside2level",
    "pad_backend",
    "pad_with_dim",
    "pad",
    "PaddingBackends",
    "pcolormesh",
    "PixelOrder",
    "reorder",
    "ring2double",
    "ring2xy",
    "to_double_pixelization",
    "to_rotated_pixelization",
    "XY",
    "xy2xy",
    "zonal_average",
]
