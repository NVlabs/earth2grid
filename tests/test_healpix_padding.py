import numpy
import torch

from earth2grid.healpix import XY, Grid
from earth2grid.healpix.padding import _xy_with_filled_tile, local2xy


def test_hpx_pad(regtest):
    order = 2
    nside = 2**order
    pad = pad_x = nside
    face = 5

    grid = Grid(order, pixel_order=XY())
    lat = torch.from_numpy(grid.lat)

    x = torch.arange(-pad_x, nside + pad_x)
    y = torch.arange(-pad, nside + pad)
    f = torch.tensor([face])

    f, y, x = torch.meshgrid(f, y, x, indexing="ij")

    x1, y1, f1 = local2xy(nside, x, y, f)

    def _to_pix(xy):
        x, y, f = xy
        return torch.where(f < 12, nside**2 * f + nside * y + x, -1)

    xy_east, xy_west = _xy_with_filled_tile(nside, x1, y1, f1)

    xy_east = _to_pix(xy_east)
    xy_west = _to_pix(xy_west)

    padded_from_west = torch.where(xy_west >= 0, lat[xy_west], 0)
    padded_from_east = torch.where(xy_east >= 0, lat[xy_east], 0)
    denom = (xy_west >= 0).int() + (xy_east >= 0).int()

    padded = (padded_from_east + padded_from_west) / denom
    numpy.savetxt(regtest, padded[0].cpu(), fmt="%.2f")
