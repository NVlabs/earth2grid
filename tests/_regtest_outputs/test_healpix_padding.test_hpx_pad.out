-78.28 -78.28 -66.44 -54.34 -78.28 -66.44 -54.34 -41.81 -30.00 -19.47 -9.59 0.00
-78.28 -66.44 -66.44 -54.34 -66.44 -54.34 -41.81 -30.00 -19.47 -9.59 0.00 9.59
-66.44 -66.44 -54.34 -54.34 -54.34 -41.81 -30.00 -19.47 -9.59 0.00 9.59 19.47
-54.34 -54.34 -54.34 -41.81 -41.81 -30.00 -19.47 -9.59 0.00 9.59 19.47 30.00
-78.28 -66.44 -54.34 -41.81 -30.00 -19.47 -9.59 0.00 9.59 19.47 30.00 41.81
-66.44 -54.34 -41.81 -30.00 -19.47 -9.59 0.00 9.59 19.47 30.00 41.81 54.34
-54.34 -41.81 -30.00 -19.47 -9.59 0.00 9.59 19.47 30.00 41.81 54.34 66.44
-41.81 -30.00 -19.47 -9.59 0.00 9.59 19.47 30.00 41.81 54.34 66.44 78.28
-30.00 -19.47 -9.59 0.00 9.59 19.47 30.00 41.81 41.81 54.34 54.34 54.34
-19.47 -9.59 0.00 9.59 19.47 30.00 41.81 54.34 54.34 54.34 66.44 66.44
-9.59 0.00 9.59 19.47 30.00 41.81 54.34 66.44 54.34 66.44 66.44 78.28
0.00 9.59 19.47 30.00 41.81 54.34 66.44 78.28 54.34 66.44 78.28 78.28
